### PR TITLE
Support LCS

### PR DESCRIPTION
**Pull request by Avital-Fine**
Add support to new [LCS](https://redis.io/commands/lcs) command, and deprecate STRALGO command

### CODE REVIEW

**Comment by Avital-Fine** [comment]
Can we deprecate strAlgo now that we have LCS? It supposes to be deprecated in Redis 7.0.0

File: src/main/java/redis/clients/jedis/commands/StringBinaryCommands.java [hunk 1]
```diff
@@ -76,5 +73,7 @@
 
   long bitop(BitOP op, byte[] destKey, byte[]... srcKeys);
 
-  LCSMatchResult strAlgoLCSKeys(final byte[] keyA, final byte[] keyB, final StrAlgoLCSParams params);
+  LCSMatchResult strAlgoLCSKeys(byte[] keyA, byte[] keyB, StrAlgoLCSParams params);
+
+  LCSMatchResult lcs(byte[] keyA, byte[] keyB, LCSParams params);
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/commands/StringBinaryCommands.java:78]: Add javadoc

File: src/main/java/redis/clients/jedis/commands/StringPipelineBinaryCommands.java [hunk 2]
```diff
@@ -5,10 +5,7 @@
 import redis.clients.jedis.Response;
 import redis.clients.jedis.args.BitCountOption;
 import redis.clients.jedis.args.BitOP;
-import redis.clients.jedis.params.BitPosParams;
-import redis.clients.jedis.params.GetExParams;
-import redis.clients.jedis.params.SetParams;
-import redis.clients.jedis.params.StrAlgoLCSParams;
+import redis.clients.jedis.params.*;
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/commands/StringPipelineBinaryCommands.java:8]: Do not use `.*` in import

File: src/main/java/redis/clients/jedis/commands/StringPipelineBinaryCommands.java [hunk 3]
```diff
@@ -77,5 +74,7 @@
 
   Response<Long> bitop(BitOP op, byte[] destKey, byte[]... srcKeys);
 
-  Response<LCSMatchResult> strAlgoLCSKeys(final byte[] keyA, final byte[] keyB, final StrAlgoLCSParams params);
+  Response<LCSMatchResult> strAlgoLCSKeys(byte[] keyA, byte[] keyB, StrAlgoLCSParams params);
+
+  Response<LCSMatchResult> lcs(byte[] keyA, byte[] keyB, LCSParams params);
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/commands/StringPipelineBinaryCommands.java:79]: Add javadoc

File: src/main/java/redis/clients/jedis/commands/StringPipelineCommands.java [hunk 4]
```diff
@@ -3,10 +3,7 @@
 import redis.clients.jedis.Response;
 import redis.clients.jedis.args.BitCountOption;
 import redis.clients.jedis.args.BitOP;
-import redis.clients.jedis.params.BitPosParams;
-import redis.clients.jedis.params.GetExParams;
-import redis.clients.jedis.params.SetParams;
-import redis.clients.jedis.params.StrAlgoLCSParams;
+import redis.clients.jedis.params.*;
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/commands/StringPipelineCommands.java:6]: Do not use `.*` in import

File: src/main/java/redis/clients/jedis/params/LCSParams.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+  @Override
+  public void addParams(CommandArguments args) {
+    if (len) {
+      args.add(LEN);
+    }
+    if (idx) {
+      args.add(IDX);
+    }
+    if (minMatchLen != null) {
```
Comment (zeekling) [inline src/main/java/redis/clients/jedis/params/LCSParams.java:65]:
```suggestion
    if (minMatchLen != null && minMatchLen >= 0) {
```

**Review by zeekling — CHANGES_REQUESTED** [review]

**Comment by sazzad16** [comment]
> Can we deprecate strAlgo now that we have LCS? It supposes to be deprecated in Redis 7.0.0

@Avital-Fine We absolutely can. You can include those in this PR or do that in a separate PR or simply create an issue :slightly_smiling_face:

**Comment by Avital-Fine** [comment]
> > Can we deprecate strAlgo now that we have LCS? It supposes to be deprecated in Redis 7.0.0
> 
> @Avital-Fine We absolutely can. You can include those in this PR or do that in a separate PR or simply create an issue slightly_smiling_face

Cool, I'm doing it in this PR

File: src/main/java/redis/clients/jedis/params/LCSParams.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+  @Override
+  public void addParams(CommandArguments args) {
+    if (len) {
+      args.add(LEN);
+    }
+    if (idx) {
+      args.add(IDX);
+    }
+    if (minMatchLen != null && minMatchLen >= 0) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/LCSParams.java:65]:
```suggestion
    if (minMatchLen != null) {
```

If anybody wants, let them watch what happens when `minMatchLen < 0`.

**Review by sazzad16 — CHANGES_REQUESTED** [review]
AFAIK, `LCS` command is not available before Redis 7. So user can't just start using it. In that case deprecation message can say,

    STRALGO LCS command will be removed from Redis 7. {@link ...} can be used instead of this method.

File: src/main/java/redis/clients/jedis/commands/StringCommands.java [hunk 7]
```diff
@@ -436,11 +437,23 @@
...
   /**
    * Calculate the longest common subsequence of keyA and keyB.
-   * @param keyA
-   * @param keyB
-   * @param params
+   * @deprecated STRALGO LCS command will be removed from Redis 7.
+   * {@link StringCommands#lcs(String, String, LCSParams) LCS} can be used instead of this method.
+   * @param keyA keyA
+   * @param keyB keyB
+   * @param params {@link StrAlgoLCSParams}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/StringCommands.java:444]:
```suggestion
   * @param keyA
   * @param keyB
   * @param params
```

File: src/main/java/redis/clients/jedis/commands/StringCommands.java [hunk 8]
```diff
@@ -436,11 +437,23 @@
...
    * @return According to StrAlgoLCSParams to decide to return content to fill LCSMatchResult.
    */
+  @Deprecated
   LCSMatchResult strAlgoLCSKeys(String keyA, String keyB, StrAlgoLCSParams params);
 
+  /**
+   * Calculate the longest common subsequence of keyA and keyB.
+   * @param keyA keyA
+   * @param keyB keyB
+   * @param params {@link LCSParams}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/StringCommands.java:454]:
```suggestion
   * @param keyA
   * @param keyB
   * @param params
```

File: src/main/java/redis/clients/jedis/commands/StringBinaryCommands.java [hunk 9]
```diff
@@ -76,6 +77,19 @@
...
+   * {@link StringBinaryCommands#lcs(byte[], byte[], LCSParams) LCS} can be used instead of this method.
+   */
+  @Deprecated
   LCSMatchResult strAlgoLCSKeys(byte[] keyA, byte[] keyB, StrAlgoLCSParams params);
 
+  /**
+   * Calculate the longest common subsequence of keyA and keyB.
+   * @param keyA keyA
+   * @param keyB keyB
+   * @param params {@link LCSParams}
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/StringBinaryCommands.java:91]:
```suggestion
   * @param keyA
   * @param keyB
   * @param params
```

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 10]
```diff
@@ -7369,11 +7384,25 @@ public LCSMatchResult strAlgoLCSKeys(final String keyA, final String keyB, final
...
   public LCSMatchResult strAlgoLCSStrings(final String strA, final String strB, final StrAlgoLCSParams params) {
     checkIsInMultiOrPipeline();
     return connection.executeCommand(commandObjects.strAlgoLCSStrings(strA, strB, params));
   }
 
+  /**
+   * Calculate the longest common subsequence of keyA and keyB.
+   * @param keyA keyA
+   * @param keyB keyB
+   * @param params the params
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:7397]:
```suggestion
   * @param keyA
   * @param keyB
   * @param params
```

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 11]
```diff
@@ -7351,12 +7363,15 @@ public long strlen(final String key) {
...
   /**
    * Calculate the longest common subsequence of keyA and keyB.
-   * @param keyA
-   * @param keyB
-   * @param params
+   * @deprecated STRALGO LCS command will be removed from Redis 7.
+   * {@link Jedis#lcs(String, String, LCSParams) LCS} can be used instead of this method.
+   * @param keyA keyA
+   * @param keyB keyB
+   * @param params the params
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:7370]:
```suggestion
   * @param keyA
   * @param keyB
   * @param params
```

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 12]
```diff
@@ -3512,17 +3512,29 @@ public long strlen(final byte[] key) {
...
+   * {@link Jedis#lcs(byte[], byte[], LCSParams) LCS} can be used instead of this method.
+   */
   @Override
+  @Deprecated
   public LCSMatchResult strAlgoLCSKeys(final byte[] keyA, final byte[] keyB, final StrAlgoLCSParams params) {
     checkIsInMultiOrPipeline();
     return connection.executeCommand(commandObjects.strAlgoLCSKeys(keyA, keyB, params));
   }
 
+  @Deprecated
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:3526]:
```suggestion
  /**
   * @deprecated STRALGO LCS command will be removed from Redis 7.
   */
  @Deprecated
```

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 13]
```diff
@@ -7369,11 +7384,25 @@ public LCSMatchResult strAlgoLCSKeys(final String keyA, final String keyB, final
    * @param params
    * @return According to StrAlgoLCSParams to decide to return content to fill LCSMatchResult.
    */
+  @Deprecated
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:7387]: `   * @deprecated STRALGO LCS command will be removed from Redis 7.`

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 14]
```diff
@@ -575,14 +575,24 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
...
 
   public final CommandObject<LCSMatchResult> strAlgoLCSKeys(byte[] keyA, byte[] keyB, StrAlgoLCSParams params) {
-    return new CommandObject<>(commandArguments(STRALGO).add(LCS).add(Keyword.KEYS)
+    return new CommandObject<>(commandArguments(STRALGO).add(Keyword.LCS).add(Keyword.KEYS)
         .key(keyA).key(keyB).addParams(params), BuilderFactory.STR_ALGO_LCS_RESULT_BUILDER);
   }
+
+  public final CommandObject<LCSMatchResult> lcs(String keyA, String keyB, LCSParams params) {
+    return new CommandObject<>(commandArguments(Command.LCS).key(keyA).key(keyB)
+            .addParams(params), BuilderFactory.STR_ALGO_LCS_RESULT_BUILDER);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/CommandObjects.java:589]:
```suggestion
    return new CommandObject<>(commandArguments(Command.LCS).key(keyA).key(keyB)
        .addParams(params), BuilderFactory.STR_ALGO_LCS_RESULT_BUILDER);
```

File: src/main/java/redis/clients/jedis/Protocol.java [hunk 15]
```diff
@@ -236,7 +236,7 @@ public static Object read(final RedisInputStream is) {
     XDEL, XTRIM, XRANGE, XREVRANGE, XREAD, XACK, XGROUP, XREADGROUP, XPENDING, XCLAIM, XAUTOCLAIM,
     XINFO, BITFIELD_RO, LPOS, ZMSCORE, BZPOPMIN, BZPOPMAX, BLMOVE, LMOVE, COPY, ROLE, FAILOVER,
     STRALGO, GEOSEARCH, GEOSEARCHSTORE, LOLWUT, REPLICAOF, ZRANGESTORE, SINTERCARD, ZINTERCARD,
-    SORT_RO, EVAL_RO, EVALSHA_RO;
+    SORT_RO, EVAL_RO, EVALSHA_RO, LCS;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Protocol.java:239]: STRALGO can be deprecated.

File: src/main/java/redis/clients/jedis/Protocol.java [hunk 16]
```diff
@@ -260,7 +260,7 @@ private Command() {
     CAT, GENPASS, LOG, INCR, SAVE, JUSTID, WITHVALUES, UNBLOCK, NOMKSTREAM, MINID, DB, ABSTTL, TO,
     TIMEOUT, ABORT, NX, XX, EX, PX, EXAT, PXAT, CH, WITHCOORD, WITHDIST, WITHHASH, STOREDIST, COPY,
     KEEPTTL, AUTH, AUTH2, INFO, CHANNELS, NUMPAT, NUMSUB, LCS, KEYS, STRINGS, FULL, NOW, VERSION,
-    ANY, FROMMEMBER, FROMLONLAT, BYRADIUS, BYBOX, BYLEX, BYSCORE, REV;
+    ANY, FROMMEMBER, FROMLONLAT, BYRADIUS, BYBOX, BYLEX, BYSCORE, REV, IDX, MINMATCHLEN, WITHMATCHLEN;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Protocol.java:263]: LCS and STRINGS can be deprecated.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Review by sazzad16 — CHANGES_REQUESTED** [review]
Hopefully last change request in this PR. Deprecate in `src/main/java/redis/clients/jedis/CommandObjects.java`

**Pull request merged**